src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.SkeletonReferenceCache.cs

```diff
@@ -266,82 +275,34 @@ private sealed class SkeletonReferenceSet
-            {
-                metadataReference = null;
-                if (!_metadataReferences.TryGetValue(properties, out var weakMetadata))
-                    return false;
-
-                // If we are pointing at a null-weak reference (not a weak reference that points to null), then we 
-                // know we failed to create the metadata the last time around, and we can shortcircuit immediately,
-                // returning null *with* success to bubble that up.
-                if (weakMetadata == null)
-                    return true;
 
-                return weakMetadata.TryGetTarget(out metadataReference);
+                // note: computing the assembly metadata is actually synchronous.  However, this ensures we don't have N
+                // threads blocking on a lazy to compute the work.  Instead, we'll only occupy one thread, while any
+                // concurrent requests asynchronously wait for that work to be one.
+                _metadata = new AsyncLazy<AssemblyMetadata?>(c => Task.FromResult(ComputeMetadata(_storage, c)), cacheResult: true);
             }
 
-            private MetadataReference? CreateReference(ImmutableArray<string> aliases, bool embedInteropTypes, DocumentationProvider documentationProvider)
+            private static AssemblyMetadata? ComputeMetadata(ITemporaryStreamStorage? storage, CancellationToken cancellationToken)
             {
-                if (_storage == null)
+                if (storage == null)
                     return null;
 
                 // first see whether we can use native memory directly.
-                var stream = _storage.ReadStream();
-                AssemblyMetadata metadata;
+                var stream = storage.ReadStream(cancellationToken);
 
                 if (stream is ISupportDirectMemoryAccess supportNativeMemory)
                 {
                     // this is unfortunate that if we give stream, compiler will just re-copy whole content to 
                     // native memory again. this is a way to get around the issue by we getting native memory ourselves and then
                     // give them pointer to the native memory. also we need to handle lifetime ourselves.
-                    metadata = AssemblyMetadata.Create(ModuleMetadata.CreateFromImage(supportNativeMemory.GetPointer(), (int)stream.Length));
+                    var metadata = AssemblyMetadata.Create(ModuleMetadata.CreateFromImage(supportNativeMemory.GetPointer(), (int)stream.Length));
 
                     // Tie lifetime of stream to metadata we created. It is important to tie this to the Metadata and not the
                     // metadata reference, as PE symbols hold onto just the Metadata. We can use Add here since we created
                     // a brand new object in AssemblyMetadata.Create above.
                     s_lifetime.Add(metadata, supportNativeMemory);
```

Comment:
>s_lifetime [](http://example.com/codeflow?start=20&length=10)

Since we now hold on the metadata in AsyncLazy, do we actually need this?

Reply:
We do yes, and it's vital (fortunately tests will crash).  The reason for this is that metadata-references are handed *out* and can effectively live *longer* than our own internal data structures.  One way to imagine this is simply that we handle out compilations through our Project API, but the user then lets go of the Solution instance (or it moves forward on the workspace).  All the existing snapshots in the Solution/Project/CompolationTracker snapshot get GC'ed and explosions happen.

Effectively, we're saying *as long as anything is referencing this metadata, this stream cannot be removed*.

--

Note: i learned this from actually trying to remove it and discovering that yes it does blow up :)

Reply:
Yeah, i see that we need to hold on the native memory object, but we don't need another weak reference for that since we already strongly reference metadata object. We can also strongly reference the stream in the same place (i.e. in `AsyncLazy<AssemblyMetadata?>`), no?

Reply:
I see, so alternatively we could implement a subtype of `PortableExecutableReference`, return an instance of it and store the memory object to it.

Reply:
I wonder if it would be best to add `ModuleMetadata.CreateFromImage()` overload that takes an `object memoryOwner`. That way whenever you create a pointer based module metadata you can just pass in the object that keeps it alive and not worry about it.

Reply:
> I see, so alternatively we could implement a subtype of PortableExecutableReference, return an instance of it and store the memory object to it.

Yes.  That would work.  But there's already a subclass being returned.  Slightly worried about having to preserve semantics with that.  However, i am very open to this as i hate CWTs and would love to get rid of that, so let me try that.

Reply:
> I wonder if it would be best to add ModuleMetadata.CreateFromImage() overload that takes an object memoryOwner. That way whenever you create a pointer based module metadata you can just pass in the object that keeps it alive and not worry about it.

That would be a way to do it :)

Reply:
ok.  i started down teh compiler side, and it was a lot of changes to the public API.  I'll do that in a followup.  FOr now though i made a custom subclass where we can hold onto the memory ourselves.